[JSC] JIT stub should hold Watchpoint and WatchpointSet

Constellation opened the pull request:
#### 159f9e46d6bb256c1de1ed39f66935ccaf987734
<pre>
[JSC] JIT stub should hold Watchpoint and WatchpointSet
<a href="https://bugs.webkit.org/show_bug.cgi?id=273454">https://bugs.webkit.org/show_bug.cgi?id=273454</a>
<a href="https://rdar.apple.com/127262594">rdar://127262594</a>

Reviewed by Keith Miller.

Previously, our approach is

1. Collecting WatchpointSet throughout IC
2. Watch them and holding watchpoints in StructureStubInfo (or its owining data structures)
3. When WatchpointSet fires, clearing StructureStubInfo.

Instead, the new approach is

1. Collecting WatchpointSet throughout IC
2. Watch them and holding watchpoints in IC JIT Stub
3. IC JIT Stub also holds WatchpointSet
4. StructureStubInfo watches (3)&apos;s WatchpointSet instead.
5. When WatchpointSet fires, invalidating (3)&apos;s IC JIT Stub&apos;s WatchpointSet.
6. Because of (5), WatchpointSet fires, and clearing StructureStubInfo.

In this way, we can have one WatchpointSet managed by IC JIT Stub. And each *potentially multiple* StructureStubInfo can monitor this condition
by inserting Watchpoint into this set. In this way, we can share one IC JIT Stub with multiple StructureStubInfo even if the IC has watchpoints.
This is the work towards Handler IC.

* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
(JSC::InlineCacheHandler::InlineCacheHandler):
(JSC::m_watchpoint):
(JSC::InlineCacheCompiler::generateImpl):
(JSC::InlineCacheCompiler::regenerate):
(JSC::m_watchpoints): Deleted.
* Source/JavaScriptCore/bytecode/InlineCacheCompiler.h:
* Source/JavaScriptCore/bytecode/StructureStubClearingWatchpoint.cpp:
(JSC::StructureStubInfoClearingWatchpoint::fireInternal):
(JSC::StructureTransitionStructureStubClearingWatchpoint::fireInternal):
(JSC::WatchpointsOnStructureStubInfo::ensureReferenceAndInstallWatchpoint):
(JSC::WatchpointsOnStructureStubInfo::ensureReferenceAndAddWatchpoint):
(JSC::AdaptiveValueStructureStubClearingWatchpoint::handleFire):
(JSC::WatchpointsOnStructureStubInfo::isValid const): Deleted.
* Source/JavaScriptCore/bytecode/StructureStubClearingWatchpoint.h:
* Source/JavaScriptCore/bytecode/Watchpoint.h:
* Source/JavaScriptCore/jit/GCAwareJITStubRoutine.cpp:
(JSC::PolymorphicAccessJITStubRoutine::PolymorphicAccessJITStubRoutine):
(JSC::PolymorphicAccessJITStubRoutine::setWatchpoints):
(JSC::PolymorphicAccessJITStubRoutine::invalidate):
* Source/JavaScriptCore/jit/GCAwareJITStubRoutine.h:
(JSC::PolymorphicAccessJITStubRoutine::watchpoints const):
(JSC::PolymorphicAccessJITStubRoutine::watchpointSet):

Canonical link: <a href="https://commits.webkit.org/278148@main">https://commits.webkit.org/278148@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1d38e18fa8da5c1319a54c08734c88bd0c5e8106

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49658 "Failed to checkout and rebase branch from PR 27908") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28946 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/52707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/52900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34965 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26563 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/52900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51758 "Failed to checkout and rebase branch from PR 27908") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/26475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/52707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/52900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/23918 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/52707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8029 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/42973 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/45826 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/52707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54477 "Built successfully") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/49153 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/24745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/20903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/47892 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/26012 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/52707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/46919 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/26860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/56640 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7140 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/25737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/11640 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/jit/GCAwareJITStubRoutine.cpp:
```diff
@@ -101,6 +109,12 @@ void PolymorphicAccessJITStubRoutine::observeZeroRefCountImpl()
 
+void PolymorphicAccessJITStubRoutine::invalidate()
+{
+    StringFireDetail detail("PolymorphicAccessJITStubRoutine gets invalidated");
```
nit: drop gets or gets -> has been

Constellation replied:
Fixed.

kmiller68 commented on Source/JavaScriptCore/bytecode/InlineCacheCompiler.h:
```diff
@@ -194,14 +193,14 @@ class InlineCacheHandler final : public RefCounted<InlineCacheHandler> {
     CodePtr<JITStubRoutinePtrTag> m_jumpTarget;
     RefPtr<PolymorphicAccessJITStubRoutine> m_stubRoutine;
-    std::unique_ptr<WatchpointsOnStructureStubInfo> m_watchpoints;
+    std::unique_ptr<StructureStubInfoClearingWatchpoint> m_watchpoint;
```
Why not have `InlineCacheHandler` subclass `StructureStubInfoClearingWatchpoint` or embed the watchpoint inline (not sure if that's possible)?

Constellation replied:
Right now, we are just using the current existing mechanism. It is nice if we can refactor after seeing all things get working well. But for now, I think it is too early to do some refactoring.

kmiller68 commented on Source/JavaScriptCore/jit/GCAwareJITStubRoutine.h:
```diff
@@ -116,6 +123,8 @@ class PolymorphicAccessJITStubRoutine : public GCAwareJITStubRoutine {
     FixedVector<RefPtr<AccessCase>> m_cases;
     FixedVector<StructureID> m_weakStructures;
     FixedVector<Identifier> m_identifiers;
+    Ref<WatchpointSet> m_watchpointSet;
```
Maybe we should have a EmbeddedWatchpointSet which is just `WatchpointSet` without the `ThreadSafeRefCounted<WatchpointSet>`?

Constellation replied:
Yeah, I think this is good extension in the later patch. For now, I'm just using the existing mechanism.

kmiller68 reviewed (approved):
> 2. Watch them and holding watchpoints in StructureStubInfo (or its owining data structures)
typo: holding -> hold and owining -> owning

> 2. Watch them and holding watchpoints in IC JIT Stub
typo: holding -> hold

> 6. Because of (6), WatchpointSet fires, and clearing StructureStubInfo.
I assume (6) should be (5)?

r=me with improvement ideas.

Constellation committed:
[JSC] JIT stub should hold Watchpoint and WatchpointSet
https://bugs.webkit.org/show_bug.cgi?id=273454
rdar://127262594

Reviewed by Keith Miller.

Previously, our approach is

1. Collecting WatchpointSet throughout IC
2. Watch them and holding watchpoints in StructureStubInfo (or its owining data structures)
3. When WatchpointSet fires, clearing StructureStubInfo.

Instead, the new approach is

1. Collecting WatchpointSet throughout IC
2. Watch them and holding watchpoints in IC JIT Stub
3. IC JIT Stub also holds WatchpointSet
4. StructureStubInfo watches (3)'s WatchpointSet instead.
5. When WatchpointSet fires, invalidating (3)'s IC JIT Stub's WatchpointSet.
6. Because of (5), WatchpointSet fires, and clearing StructureStubInfo.

In this way, we can have one WatchpointSet managed by IC JIT Stub. And each *potentially multiple* StructureStubInfo can monitor this condition
by inserting Watchpoint into this set. In this way, we can share one IC JIT Stub with multiple StructureStubInfo even if the IC has watchpoints.
This is the work towards Handler IC.

* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
(JSC::InlineCacheHandler::InlineCacheHandler):
(JSC::m_watchpoint):
(JSC::InlineCacheCompiler::generateImpl):
(JSC::InlineCacheCompiler::regenerate):
(JSC::m_watchpoints): Deleted.
* Source/JavaScriptCore/bytecode/InlineCacheCompiler.h:
* Source/JavaScriptCore/bytecode/StructureStubClearingWatchpoint.cpp:
(JSC::StructureStubInfoClearingWatchpoint::fireInternal):
(JSC::StructureTransitionStructureStubClearingWatchpoint::fireInternal):
(JSC::WatchpointsOnStructureStubInfo::ensureReferenceAndInstallWatchpoint):
(JSC::WatchpointsOnStructureStubInfo::ensureReferenceAndAddWatchpoint):
(JSC::AdaptiveValueStructureStubClearingWatchpoint::handleFire):
(JSC::WatchpointsOnStructureStubInfo::isValid const): Deleted.
* Source/JavaScriptCore/bytecode/StructureStubClearingWatchpoint.h:
* Source/JavaScriptCore/bytecode/Watchpoint.h:
* Source/JavaScriptCore/jit/GCAwareJITStubRoutine.cpp:
(JSC::PolymorphicAccessJITStubRoutine::PolymorphicAccessJITStubRoutine):
(JSC::PolymorphicAccessJITStubRoutine::setWatchpoints):
(JSC::PolymorphicAccessJITStubRoutine::invalidate):
* Source/JavaScriptCore/jit/GCAwareJITStubRoutine.h:
(JSC::PolymorphicAccessJITStubRoutine::watchpoints const):
(JSC::PolymorphicAccessJITStubRoutine::watchpointSet):

Canonical link: https://commits.webkit.org/278148@main
